**<h1 style="color:#037b90">MODULE 2: INTRODUCTION TO PANDAS</h1>**

**<h3 style="color:#037b90">Module overview</h3>**

In this module you will get an in-depth introduction to Pandas library, a powerful tool for data manipulation and analysis in Python. you will be able to learn about the core data structures in Pandas, including Series and DataFrames, and explore essential functions for data manipulation such as reindexing, dropping entries, selection and filtering, and sorting. Additionally, you will learn about the techniques for summarizing data and computing descriptive statistics. This module will equip you with the skills to perform comprehensive data analysis tasks using pandas.

**<h3 style="color:#037b90">Learning Outcomes</h3>**

By the end of this module you should be able to;

1. Describe the primary data structures in Pandas
2.  Explain the processes and significance of reindexing, dropping entries, selection and filtering, and sorting and ranking in Pandas
3. Apply Pandas functions to reindex a DataFrame, drop specific entries, and filter data based on given criteria
4. Analyze a dataset to compute and interpret descriptive statistics, such as mean, median, and standard deviation using Pandas.

**<h3 style="color:#037b90">CONTENT</h3>**

Introduction to Pandas: Pandas data structures(Series and dataframes), Essential functions(Reindexing, dropping entries, selection and filtering, sorting and ranking); Summarizing and computing descriptive statistics

**What is pandas**

Pandas is a python library that is used for data analysis. Pandas stands for panel data. The name is derived from the fact that the library displays data in a tabular manner like excel.

Pandas has two data structures that we will be looking at in this section namely series and dataframe. Before we start looking at how the two data structures work, we will start by making sure that we have pandas in our workspace. If you have not installed pandas you can use the following command to do so

    pip install pandas

After you have installed the pandas module you can now import it to your workspace using the import keyword as follows

    import pandas as pd
    
We are using the alias pd to make it easier to reference the pandas library

**Series**: A series is basically a one dimensional array that is able to hold any python data type. It has axis labels which are used as index.

**Creating a series**

The most basic way of creating a series is using the pandas Series method. The series can be created from a list, a dictionary, an ndarray or a scalar value like 5



In [1]:
#creating a series from a list
import pandas as pd
data=[34,56,77,89,34,25,67,89,86,12,59,87]
pd.Series(data)

0     34
1     56
2     77
3     89
4     34
5     25
6     67
7     89
8     86
9     12
10    59
11    87
dtype: int64

In [58]:
#Creating a series from a dictionary
dic={"A":23, "b":40, "c":67, "d":23}
dic_seris=pd.Series(dic)
print(dic_seris)

A    23
b    40
c    67
d    23
dtype: int64


**Attributes of the series data structure**

**index**: Indicates the of the series. It will show all the indexes in the series

**values**: Indicates the values in the series as a list of values

**dtype**: Indicates the data type of the data stored in the series

**shape**: Indicates the shape of the data stored in the tupple

**ndim**: Indicates the dimensions in the data in the series

**empty**: Indicates if the series is empty or not. If the series is empty, the empty attribute returns True if the series is not empty, the empty attribute returns true

In [65]:
print(dic_seris.index)
print(dic_seris.values)
print(dic_seris.dtype)
print(dic_seris.shape)
print(dic_seris.ndim)
print(dic_seris.empty)

Index(['A', 'b', 'c', 'd'], dtype='object')
[23 40 67 23]
int64
(4,)
1
False


**Accessing series element**

We can access series elements using the index. For example in our created series called dic_seris, the index will be the letters 'A', 'B', 'C'. To access the elements we use indexing as you did while accessing the list, in this case the index are the letters as shown below

In [66]:
dic_seris['A']

23

We are also be able to access the elements of the series using the number indexing as shown below

In [68]:
dic_seris[0:4]

A    23
b    40
c    67
d    23
dtype: int64

**Dataframes**

The dataframe is the other data structure found in pandas. The dataframe arranges data in a table like format.It presents data in a two dimensional array of rows and columns We can create a dataframe from values, from existing data structures or from a file read using pandas. You can use the DataFrame() function to create a dataframe.

Let us create a dictionary and convert it to a dataframe


In [5]:
dicti={"a":[40,40], "b":[56,89],"c":[34,56], "d":[66,77]}
data=pd.DataFrame(data=dicti)

In [6]:
data

a   b   c   d
0  40  56  34  66
1  40  89  56  77

You can use the DataFrame() fuction to create a dataframe from a numpy array as follows;

In [2]:
import numpy as np
import pandas as pd
data = np.array([(22, 56, 89), (41, 59, 63), (70, 80, 90)],
                dtype=[("a", "i4"), ("b", "i4"), ("c", "i4")])
dtarra=pd.DataFrame(data)
dtarra

a   b   c
0  22  56  89
1  41  59  63
2  70  80  90

Pandas has other functions for reading data from files such as csv, excel, stata. Most organizations do have data in form of excel sheet or the CSV file format. In this section, you will learn how to read data from an excel file and csv file formats

The methods that exist in pandas for reading excel and csv file are read_excel() and read_csv() respectivelly. Lets look at how to use this with an example.

We will create a variable called data to store our data that is read from the csv file



In [2]:
#reading a csv file 
import pandas as pd
data=pd.read_csv("Student_Performance.csv")

Now that we have read our data, we can peek through the data to see how it looks like using the functions pandas as provided for us.

**head()**: The head() functions displays the first five rows of the data. If you want to display more than five rows, we can specifify the number inside the opening and closing braces. Using the data we read from above, we can explore it using the head() function as follows

In [12]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
print(data.head())

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   
2              8               51                        Yes            7   
3              5               52                        Yes            5   
4              7               75                         No            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1                 91  
1                                 2                 65  
2                                 2                 45  
3                                 2                 36  
4                                 5                 66

In [13]:
#displays 2 rows
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
print(data.head(2))

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   

   Sample Question Papers Practiced  Performance Index  
0                                 1                 91  
1                                 2                 65

If you want to see the last rows of the dataset, we can use the tail() function as follows

In [14]:
#displaying the last five elements
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.tail()

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
9995              1               49                        Yes            4   
9996              7               64                        Yes            8   
9997              6               83                        Yes            8   
9998              9               97                        Yes            7   
9999              7               74                         No            8   

      Sample Question Papers Practiced  Performance Index  
9995                                 2                 23  
9996                                 5                 58  
9997                                 5                 74  
9998                                 0                 95  
9999                                 1                 64

In [15]:
#displaying the last two rows of the dataset
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.tail(2)

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
9998              9               97                        Yes            7   
9999              7               74                         No            8   

      Sample Question Papers Practiced  Performance Index  
9998                                 0                 95  
9999                                 1                 64

If we want to see only a specific column in the dataset, we can reference the column name using the square paranthesis as we used to reference the elements in the list.  

In [16]:
#referencing a specific column
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data['Hours Studied']

0       7
1       4
2       8
3       5
4       7
       ..
9995    1
9996    7
9997    6
9998    9
9999    7
Name: Hours Studied, Length: 10000, dtype: int64

We can also check the various attributes of the data cointained in the dataframe. For example if you want to see the number of columns and rows in the dataset, we can use the shape attribute to do so as follows

In [18]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.shape

(10000, 6)

The shape attribute returns a tupple with two values. The first value is the number of rows and the second value is the number of columns. Our dataset in this case therefore has 10,000 rows and 6 columns. We can also extract the column names from the dataset by using the columns attribute as follows

In [19]:
#accessing the column names from the dataset
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.columns

Index(['Hours Studied', 'Previous Scores', 'Extracurricular Activities',
       'Sleep Hours', 'Sample Question Papers Practiced', 'Performance Index'],
      dtype='object')

The columns attribute returns all the column names in the dataset.You can also view the information about our dataset using the info() method

In [24]:
#getting the datatypes of the columns in the dataset
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   Hours Studied                     10000 non-null  int64 
 1   Previous Scores                   10000 non-null  int64 
 2   Extracurricular Activities        10000 non-null  object
 3   Sleep Hours                       10000 non-null  int64 
 4   Sample Question Papers Practiced  10000 non-null  int64 
 5   Performance Index                 10000 non-null  int64 
dtypes: int64(5), object(1)
memory usage: 468.9+ KB


From the info() method we can be able to see the datatypes of the variables, the number of non null values for each columns and the number of entries in the dataset.

**Functions for manipulating data in pandas**

The are various functions in pandas that help us in either getting information about our dataset, or manipulating our dataset. In this section, we are going to look at some of this functions in this section.

**isnull()**: The isnull() function is used to check if there are any null values in the dataset. We can use the isnull() function together with the sum() function to calculate the total null values in the data.

In [26]:
#Counting the null values in the dataframe
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.isnull().sum()

Hours Studied                       0
Previous Scores                     0
Extracurricular Activities          0
Sleep Hours                         0
Sample Question Papers Practiced    0
Performance Index                   0
dtype: int64

From the results above, we can see that we all our columns do not cointain the null values.

**<h4>DESCRIPTIVE STATISTICS</h4>**

**describe()**: The describe() function culculates basic statistical values for the numeric columns in the dataset. Using the describe() function, we can be able to see the total count for all the numeraical columns, the mean, the satndard deviation, the minimum values, the quartiles values and the maximum value.

We also have functions for culculating each of the descriptive statistics for the whole dataframe and for particular columns

In [27]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.describe()

Hours Studied  Previous Scores   Sleep Hours  \
count   10000.000000     10000.000000  10000.000000   
mean        4.992900        69.445700      6.530600   
std         2.589309        17.343152      1.695863   
min         1.000000        40.000000      4.000000   
25%         3.000000        54.000000      5.000000   
50%         5.000000        69.000000      7.000000   
75%         7.000000        85.000000      8.000000   
max         9.000000        99.000000      9.000000   

       Sample Question Papers Practiced  Performance Index  
count                      10000.000000       10000.000000  
mean                           4.583300          55.224800  
std                            2.867348          19.212558  
min                            0.000000          10.000000  
25%                            2.000000          40.000000  
50%                            5.000000          55.000000  
75%                            7.000000          71.000000  
max                            9.000000         100.000000

In [37]:
#Get the mean for the numerical columns
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.mean()

Hours Studied                        4.9929
Previous Scores                     69.4457
Sleep Hours                          6.5306
Sample Question Papers Practiced     4.5833
Performance Index                   55.2248
dtype: float64

In [38]:
#gets the sum of the various numerical columns in the dataset
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.sum()

Hours Studied                        49929
Previous Scores                     694457
Sleep Hours                          65306
Sample Question Papers Practiced     45833
Performance Index                   552248
dtype: int64

In [39]:
#Gets the mode for each colum
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.mode()

Hours Studied  Previous Scores  Sleep Hours  \
0              1               54            8   

   Sample Question Papers Practiced  Performance Index  
0                                 6                 67

In [40]:
#gets the standard deviation for each column
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.std()

Hours Studied                        2.589309
Previous Scores                     17.343152
Sleep Hours                          1.695863
Sample Question Papers Practiced     2.867348
Performance Index                   19.212558
dtype: float64

In [44]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data["Hours Studied"].sum()

49929

**drop()**: There are some instances where you want to remove a column from a dataframe. The drop function helps us achieve this. For example we can remove the "Extracurricular Activities" activities from our dataframe as follows. The drop() method expects the column name to be dropped, the inplace attribute which determines if the modifiation will be done on the original dataframe and the axis indicates whether row or column. Axis 1 is for column and 0 is for row

In [31]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.drop(columns="Extracurricular Activities",inplace=True, axis=1)
data.head()

From the results, we can see that we no longer have the "Extracurricular Activities" column

**value_counts()**: We can also count the specific values in a culumn using the value_counts() function as follows

In [33]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data["Sleep Hours"].value_counts()

Sleep Hours
8    1804
7    1676
6    1673
9    1622
4    1619
5    1606
Name: count, dtype: int64

From the results of counting the "sleep hours" column, we can see the number of people who have slept for the various hours e.g 8 hours has 1804.

**sample()**: The sample() function helps us get a sample values from our dataframe. When using the sample, we can specify the size of our sample as the parameter.

In [36]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
sample_data=data.sample(50)
print(sample_data.shape)
print(sample_data.head())

(50, 5)

**<h4>Reindexing </h4>**

If you look at our dataset, we can see that pandas provides for us an automated index. In some cases you might want to create your own indexes, in this case you can use the reindexing functions in pandas. For example if we wanted you wanted to remove the index in the dataset we read in the previsous setion, we can use the pandas reindex()function. We pass to the reindex function the new index we expect to create as our new index


In [16]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
new_index=(range(20,10020))
data.reindex(new_index)

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
20               1.0             99.0                        Yes          4.0   
21               6.0             96.0                         No          9.0   
22               9.0             74.0                        Yes          7.0   
23               1.0             85.0                         No          5.0   
24               3.0             61.0                         No          6.0   
...              ...              ...                        ...          ...   
10015            NaN              NaN                        NaN          NaN   
10016            NaN              NaN                        NaN          NaN   
10017            NaN              NaN                        NaN          NaN   
10018            NaN              NaN                        NaN          NaN   
10019            NaN              NaN                        NaN          NaN   

       Sample Question Papers Practiced  Performance Index  
20                                  3.0               71.0  
21                                  0.0               85.0  
22                                  6.0               73.0  
23                                  6.0               57.0  
24                                  3.0               35.0  
...                                 ...                ...  
10015                               NaN                NaN  
10016                               NaN                NaN  
10017                               NaN                NaN  
10018                               NaN                NaN  
10019                               NaN                NaN  

[10000 rows x 6 columns]

**set_index()**: You can use the set index to set an existing column to be the index

In [17]:
import pandas as pd
data=pd.read_csv("Student_Performance.csv")
data.set_index('Hours Studied')

Previous Scores Extracurricular Activities  Sleep Hours  \
Hours Studied                                                            
7                           99                        Yes            9   
4                           82                         No            4   
8                           51                        Yes            7   
5                           52                        Yes            5   
7                           75                         No            8   
...                        ...                        ...          ...   
1                           49                        Yes            4   
7                           64                        Yes            8   
6                           83                        Yes            8   
9                           97                        Yes            7   
7                           74                         No            8   

               Sample Question Papers Practiced  Performance Index  
Hours Studied                                                       
7                                             1                 91  
4                                             2                 65  
8                                             2                 45  
5                                             2                 36  
7                                             5                 66  
...                                         ...                ...  
1                                             2                 23  
7                                             5                 58  
6                                             5                 74  
9                                             0                 95  
7                                             1                 64  

[10000 rows x 5 columns]

**<h4>Selection Filtering</h4>**

In most cases, you might find your dataset has alot of columns and you might want to be able to see a portion of the dataset based on the need and some creteria. To do that you will need to use the selection and filtering creteria provided in pandas.

**Selecting multiple columns**

We saw earlier that we can select a single column using the dataframe name and the column name in the square brackets. To select multiple columns, we will need to pass the column name as a list in the square brackets. To do this lets read another dataset and extract a portion of it.

In [22]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
new_dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

As we can see our data has alot of columns, lets say we want to see only the Glucose and BloodPressure columns. We can do so as follows

In [24]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
new_dataset.head()
new_dataset_select=new_dataset[['Glucose','BloodPressure']]
print(new_dataset_select.head())

**<h4>Filtering:</h4>** 

Lets say that we wanted to only see the data where blood pressure is below 50. We can select only those row that have blood pressure below 50 as follows

In [27]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
bpbelow50=new_dataset[new_dataset['BloodPressure']<50]

We can look at the shape of our data to be able to see how many rows have in our new dataset

In [28]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
bpbelow50=new_dataset[new_dataset['BloodPressure']<50]
bpbelow50.shape

(51, 9)

You can see that the rows that have bp below 50 are 51. We can also select the rows that only have an outcome of 1 as follows

In [30]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
outcome1=new_dataset[new_dataset['Outcome']==1]
print(outcome1.shape)

(268, 9)

From our dataset, we can see that we have 268 records where the outcome was equals to 1.

We can also view specific rows and specific columns. We can access those using the iloc function in pandas. For example if we were interested only in the first 10 rows and the first three columns we will write

In [32]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
print(new_dataset.iloc[1:10, 1:4])

Glucose  BloodPressure  SkinThickness
1       85             66             29
2      183             64              0
3       89             66             23
4      137             40             35
5      116             74              0
6       78             50             32
7      115              0              0
8      197             70             45
9      125             96              0

If we were intereted in all the rows and only the first three columns, we will write

In [33]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
print(new_dataset.iloc[:,1:4])

Glucose  BloodPressure  SkinThickness
0        148             72             35
1         85             66             29
2        183             64              0
3         89             66             23
4        137             40             35
..       ...            ...            ...
763      101             76             48
764      122             70             27
765      121             72             23
766      126             60              0
767       93             70             31

[768 rows x 3 columns]

**<h4>Sorting</h4>**

We can sort values in the pandas dataframe using the pandas sort_values function.

The sort_values function expects the following parameters

**by:** Indicates the columnn the data will be sorted by

**axis:** The axis to be sorted 0 to sort by index, 1 or columns. The dafault is usually 0

**Ascending:** if set to True, it will sort the data in the ascending order. If set to false, it will sort the data in descending order.

**inplace:** If set to true, it will modify the dataset if set to false it will create a copy of the dataset

**na_position**: Sets where to put the null value. If the option is equal to first, it will put the null values at the top of the dataframe. If the position is set to last, it will put the null valus at the end of the dataframe.

**kind:** This specifies the type of the sorting algorithm to be used. The options are quicksort, mergesort, heapsort, stable

Lets us demostrate this using the dataset we read in the previous section. We will sort it by the age column as follows

In [34]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
new_dataset.sort_values(by='Age',inplace=True, kind="quicksort", na_position='last', ascending=True)
new_dataset.head()

If we look at the data set we can see that the columns have been sorted according to the age with 21 being the lowest age. We can peek at the tail to see the highest age. As shown in the result the highest age is 81

In [36]:
import pandas as pd
new_dataset=pd.read_csv("diabetes.csv")
new_dataset.sort_values(by='Age',inplace=True, kind="quicksort", na_position='last', ascending=True)
new_dataset.tail()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
123            5      132             80              0        0  26.8   
684            5      136             82              0        0   0.0   
666            4      145             82             18        0  32.5   
453            2      119              0              0        0  19.6   
459            9      134             74             33       60  25.9   

     DiabetesPedigreeFunction  Age  Outcome  
123                     0.186   69        0  
684                     0.640   69        0  
666                     0.235   70        1  
453                     0.832   72        0  
459                     0.460   81        0